mirrord-layer/src/go/hooks.rs

```diff
@@ -0,0 +1,423 @@
+            "mov    r8, QWORD PTR [rdi+0x30]",
+            "mov    rsi, QWORD PTR [r8+0x50]",
+            "cmp    rdi,rsi",
+            "je     2f",
+            "mov    rsi, QWORD PTR [r8]",
+            "cmp    rdi, rsi",
+            "je     2f",
+            "call   go_systemstack_switch",
+            "mov    QWORD PTR fs:[0xfffffff8], rsi",
+            "mov    rsp, QWORD PTR [rsi+0x38]",
+            "sub    rsp, 0x40",
+            "and    rsp, 0xfffffffffffffff0",
+            "mov    QWORD PTR [rsp+0x30],rdi",
+            "mov    rdi, QWORD PTR [rdi+0x8]",
+            "sub    rdi, rdx",
+            "mov    QWORD PTR [rsp+0x28], rdi",
+            "mov    rsi, rbx",
+            "mov    rdx, r10",
+            "mov    rdi, rax",
+            "call   c_abi_syscall_handler",
+            "mov    rdi, QWORD PTR [rsp+0x30]",
+            "mov    rsi, QWORD PTR [rdi+0x8]",
+            "sub    rsi, QWORD PTR [rsp+0x28]",
+            "mov    QWORD PTR fs:0xfffffff8, rdi",
+            "mov    rsp, rsi",
+            "cmp    rax, -0xfff",
+            "jbe    3f",
+            "mov    QWORD PTR [rsp+0x28], -0x1",
+            "mov    QWORD PTR [rsp+0x30], 0x0",
+            "neg    rax",
+            "mov    QWORD PTR [rsp+0x38], rax",
+            "xorps  xmm15, xmm15",
+            "mov    r14, QWORD PTR FS:[0xfffffff8]",
+            "ret",
+            "2:",
+            "sub    rsp, 0x40",
+            "and    rsp, 0xfffffffffffffff0",
+            "mov    QWORD PTR [rsp+0x30], 0x0",
+            "mov    QWORD PTR [rsp+0x28], rdx",
+            "mov    rsi, rbx",
+            "mov    rdx, r10",
+            "mov    rdi, rax",
+            "mov    r8, r11",
+            "mov    r9, r12",
+            "mov    QWORD PTR [rsp], r13",
+            "call   c_abi_syscall6_handler",
+            "mov    rsi, QWORD PTR [rsp+0x28]",
+            "mov    rsp, rsi",
+            "cmp    rax, -0xfff",
+            "jbe    3f",
+            "mov    QWORD PTR [rsp+0x28], -0x1",
+            "mov    QWORD PTR [rsp+0x30], 0x0",
+            "neg    rax",
+            "mov    QWORD PTR [rsp+0x38], rax",
+            "xorps  xmm15, xmm15",
+            "mov    r14, QWORD PTR FS:[0xfffffff8]",
+            "ret",
+            "3:",
+            "mov    QWORD PTR [rsp+0x28], rax",
+            "mov    QWORD PTR [rsp+0x30], 0x0",
+            "mov    QWORD PTR [rsp+0x38], 0x0",
+            "xorps  xmm15,xmm15",
+            "mov    r14, QWORD PTR FS:[0xfffffff8]",
+            "ret",
+            options(noreturn)
+        );
+    }
+
+    /// [Naked function] maps to gasave_systemstack_switch, called by asmcgocall.abi0
+    #[no_mangle]
+    #[naked]
+    unsafe extern "C" fn go_systemstack_switch() {
+        asm!(
+            // 0xdd9 - is the value taken from objdump, ghidra maps it to
+            // `runtime.systemstack_switch.abi0`, which just returns.
+            // TODO: fix the address so the symbol is loaded correctly
```

Comment:
I don't think it's needed - if I remember correctly, the systemstack_switch exists just to be a marker